text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,1062 @@
+      ∀i. R,r.. ⊢ Ti: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> fn(T1..Tn) -> T0
+
+    OutlivesTraitRef:
+      ∀i. R,r.. ⊢ Pi: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> TraitId<P0..Pn>: 'a      
+      
+#### Outlives for lifetimes
+
+The outlives relation for lifetimes depends on whether the lifetime in
+question was bound within a type or not. In the usual case, we decide
+the relationship between two lifetimes by consulting the environment,
+or using the reflexive property. Lifetimes representing scopes within
+the current fn have a relationship derived from the code itself, while
+lifetime parameters have relationships defined by where-clauses and
+implied bounds.
+
+    OutlivesRegionEnv:
+      'x ∉ R               // not a bound region
+      ('x: 'a) in Env      // derivable from where-clauses etc
+      --------------------------------------------------
+      R ⊢ 'x: 'a
+
+    OutlivesRegionReflexive:
+      --------------------------------------------------
+      R ⊢ 'a: 'a
+
+For higher-ranked lifetimes, we simply ignore the relation, since the
+lifetime is not yet known. This means for example that `for<'a> fn(&'a
+i32): 'x` holds, even though we do not yet know what region `'a` is
+(and in fact it may be instantiated many times with different values
+on each call to the fn).
+      
+    OutlivesRegionBound:
+      'x ∈ R               // bound region
+      --------------------------------------------------
+      R ⊢ 'x: 'a
+
+#### Outlives for type parameters
+
+For type parameters, the only way to draw "outlives" conclusions is to
+find information in the environment (which is being threaded
+implicitly here, since it is never modified). In terms of a Rust
+program, this means both explicit where-clauses and implied bounds
+derived from the signature (discussed below).
+
+    OutlivesTypeParameterEnv:
+      X: 'a in Env
+      --------------------------------------------------
+      R ⊢ X: 'a
+
+
+#### Outlives for projections
+
+Projections have the most possibilities. First, we may find
+information in the in-scope where clauses, as with type parameters,
+but we can also consult the trait definition to find bounds (consider
+an associated type declared like `type Foo: 'static`). These rule only
+apply if there are no higher-ranked lifetimes in the projection; for
+simplicity's sake, we encode that by requiring an empty list of
+higher-ranked lifetimes. (This is somewhat stricter than necessary,
+but reflects the behavior of my prototype implementation.)
+
+    OutlivesProjectionEnv:
+      <P0 as Trait<P1..Pn>>::Id: 'a in Env
+      --------------------------------------------------
+      <> ⊢ <P0 as Trait<P1..Pn>>::Id: 'a
+
+    OutlivesProjectionTraitDef:
+      WC = [Xi => Pi] WhereClauses(Trait) 
+      <P0 as Trait<P1..Pn>>::Id: 'a in WC
+      --------------------------------------------------
+      <> ⊢ <P0 as Trait<P1..Pn>>::Id: 'a
```

Comment:
@RalfJung 

> Like, say, the trait definition of T<'a, 'b> has "type Item: 'a", and then in a context where "'a: 'c" is in the environment, we need to know whether "T<'a, 'b>::Item: 'c". Neither of these two inference rules alone is good enough to capture this case, I think.

Yes, that can happen. There is a transitive rule missing, I guess. It's also true that I've tweaked the impl somewhat in the meantime to make it more expressive. I should probably just write out the rules here to be more flexible, even if the current impl doesn't always find a correct solution.